tutorials/pyroot/pyroot002_pythonizationDecorator.py

```diff
@@ -0,0 +1,141 @@
+# We will now use the `MyClass` class for the first time from Python: we will
+# create a new object of that class. At this moment, the pythonizor will
+# execute and modify the class - pythonizors are always lazily run when a given
+# class is used for the first time from a Python script.
+my_object = ROOT.MyClass()
+
+# Since the pythonizor already executed, we should now see the new behaviour.
+# For that purpose, let's print `my_object` (should show "This is a MyClass
+# object").
+print(my_object)
+
+# The previous example is just a simple one, but there are many ways in which a
+# class can be pythonized. Typical examples are the redefinition of dunder
+# methods (e.g. `__iter__` and `__next__` to make your objects iterable from
+# Python). If you need some inspiration, many ROOT classes are pythonized in
+# way we just saw; their pythonizations can be seen at:
+# https://github.com/root-project/root/tree/master/bindings/pyroot/pythonizations/python/ROOT/pythonization
+
+# The @pythonization decorator offers a few more options when it comes to
+# matching classes that you want to pythonize. We saw that we can match a
+# single class, but we can also specify a list of classes to pythonize.
+#
+# The following code defines a couple of new classes:
+ROOT.gInterpreter.Declare('''
+class Class1 {};
+
+namespace NS {
+    class Class2 {};
+}
+''')
+
+# A pythonizor that matches both classes would look like this:
+@pythonization(['Class1', 'NS::Class2'])
+def pythonize_two_classes(klass):
+    klass.new_attribute = 1
+
+# Both classes will have the new attribute:
+o1 = ROOT.Class1()
+o2 = ROOT.NS.Class2()
+print("Printing new attribute")
+for o in o1, o2:
+    print(o.new_attribute)
+
+# In addition, @pythonization also accepts prefixes of classes in a certain
+# namespace in order to match multiple classes in that namespace. In order to
+# signal that what we provide to @pythonization is a prefix, we need to set the
+# `is_prefix` argument to `True` (default is `False`).
+#
+# A common case where matching prefixes is useful is when we have a templated
+# class and we want to pythonize all possible instantiations of that template.
+# For example, we can pythonize the `std::vector` (templated) class like so:
+@pythonization('std::vector<', is_prefix=True)
+def vector_pythonizor(klass):
+    # first_elem returns the first element of the vector if it exists
+    klass.first_elem = lambda v : v[0] if v else None
+
+# Since we defined a prefix to do the match, the pythonization will be applied
+# both if we instantiate e.g. a vector of integers and a vector of doubles.
+v_int = ROOT.std.vector['int']([1,2,3])
+v_double = ROOT.std.vector['double']([4.,5.,6.])
+print("First element of integer vector: {}".format(v_int.first_elem()))
+print("First element of double vector: {}".format(v_double.first_elem()))
+
+# It is important to note that prefixes are applied to a certain namespace.
```

Comment:
Thanks, this is an important point. What I meant is that `NS1::` matches any class inside `NS1`, for example `NS1::MyClass`, but it does not match e.g. `NS1::NS2::MyClass`. So it's more "prefixes must contain the full namespace of the classes to match"?

Reply:
why does `NS1::` not match `NS1::NS2::MyClass` as a prefix? this surprising enough that it might warrant a change of name for the parameter (it's not just a string prefix if it understands namespaces and ignores nested namespaces) or separate parameters for namespaces and class name prefix. or a giant warning :)

Reply:
Indeed, and that's why I wanted a review of the tutorial :)

Nested namespaces are not matched because a pythonizor is registered in cppyy for a given namespace (e.g. `NS1::NS2`). When a class in `NS1::NS2` is used for the first time, the pythonizors in `NS1::NS2` are attempted - but not the ones in e.g. `NS1::NS2::NS3`. When a user registers a pythonizor for `NS1::NS2`, I have no means to know what nested namespaces might exist, so I can't register the pythonizor for those - and as a consequence any new class in those nested namespaces will not trigger the pythonizor.

So you would be in favour of e.g. `@pythonization(class_prefix='SomePrefix', ns='NS1::NS2')` ?